File: Sharebook/Client/src/app/core/Core/DTO/UserDTO.ts
import IEnumerable from "../../../c#-objects/TypeScript.NET-Core/packages/Core/source/Collections/Enumeration/IEnumerable";
import {IUserEntities} from "../../Public/Models/Entities/IUserEntities";
import {UserIdentifierDTO} from "./UserIdentifierDTO";
import {IUserDTO} from "../../Public/Models/Interfaces/DTO/IUserDTO";
import {ITweetDTO} from "../../Public/Models/Interfaces/DTO/ITweetDTO";
import DateTime from "../../../c#-objects/TypeScript.NET-Core/packages/Core/source/Time/DateTime";

export class UserDTO extends UserIdentifierDTO implements IUserDTO {
  // Verify : ProfileImageTile

  // [JsonProperty("name")]
  public name: string;

  // [JsonProperty("status")]
  // [JsonConverter(typeof(JsonPropertyConverterRepository))]
  public status: ITweetDTO;

  // [JsonProperty("description")]
  public description: string;

  // [JsonProperty("created_at")]
  // [JsonConverter(typeof(JsonTwitterDateTimeOffsetConverter))]
  public createdAt: DateTime;  // DateTimeOffset;

  // [JsonProperty("location")]
  public location: string;

  // [JsonProperty("geo_enabled")]
  public geoEnabled?: boolean;

  // [JsonProperty("url")]
  public url: string;

  // [JsonProperty("email")]
  public email: string;

  // [JsonProperty("statuses_count")]
  public statusesCount: number;

  // [JsonProperty("followers_count")]
  public followersCount: number;

  // [JsonProperty("friends_count")]
  public friendsCount: number;

  // [JsonProperty("following")]
  public following?: boolean;

  // [JsonProperty("protected")]
  public protected: boolean;

  // [JsonProperty("verified")]
  public verified: boolean;

  // [JsonProperty("entities")]
  // [JsonConverter(typeof(JsonPropertyConverterRepository))]
  public entities: IUserEntities;

  // [JsonProperty("notifications")]
  public notifications?: boolean;

  // [JsonProperty("profile_image_url")]
  public profileImageUrlHttp: string;

  // [JsonProperty("profile_image_url_https")]
  public profileImageUrl: string;

  // [JsonProperty("follow_request_sent")]
    public followRequestSent?: boolean;

  // [JsonProperty("default_profile")]
  public defaultProfile: boolean;

  // [JsonProperty("default_profile_image")]
  public defaultProfileImage: boolean;

  // [JsonProperty("favourites_count")]
  public favoritesCount?: number;

  // [JsonProperty("listed_count")]
  public listedCount: number;

  // [JsonProperty("profile_sidebar_fill_color")]
  public profileSidebarFillColor: string;

  // [JsonProperty("profile_sidebar_border_color")]
  public profileSidebarBorderColor: string;

  // [JsonProperty("profile_background_tile")]
  public profileBackgroundTile: boolean;

  // [JsonProperty("profile_background_color")]
  public profileBackgroundColor: string;

  // [JsonProperty("profile_background_image_url")]
  public profileBackgroundImageUrl: string;

  // [JsonProperty("profile_background_image_url_https")]
  public profileBackgroundImageUrlHttps: string;

  // [JsonProperty("profile_banner_url")]
  public profileBannerURL: string;

  // [JsonProperty("profile_text_color")]
  public profileTextColor: string;

  // [JsonProperty("profile_link_color")]
  public profileLinkColor: string;

  // [JsonProperty("profile_use_background_image")]
  public profileUseBackgroundImage: boolean;

  // [JsonProperty("is_translator")]
  public isTranslator?: boolean;

  // [JsonProperty("contributors_enabled")]
  public contributorsEnabled?: boolean;

  // [JsonProperty("utc_offset")]
  public utcOffset?: number;

  // [JsonProperty("time_zone")]
  public timeZone: string;

  // [JsonProperty("withheld_in_countries")]
  public withheldInCountries: IEnumerable<string>;

  // [JsonProperty("withheld_scope")]
  public withheldScope: string;

  public toString(): string {
    return super.screenName;
  }
}

File: Server/src/Web/Chessbook.Web.Models/Users/UserDTO.cs
namespace Chessbook.Web.Models
{
    using Newtonsoft.Json;
    using Chessbook.Data.Models;
    using Chessbook.Services.Mapping;
    using Chessbook.Web.Models.Outputs.Posts;
    using Chessbook.Web.Models.Users;
    using System;
    using System.Collections.Generic;

    public class UserDTO : UserIdentifierDTO, IMapFrom<User>
    {
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string Login { get; set; }
        public string Email { get; set; }
        public int? Age { get; set; }

        public SettingsDTO Settings { get; set; }

        // Verify : ProfileImageTile

        [JsonProperty("name")]
        public string Name { get; set; }

        //[JsonProperty("status")]
        //public PostDTO Status { get; set; }

        [JsonProperty("description")]
        public string Description { get; set; }

        [JsonProperty("created_at")]
        public DateTime CreatedOn { get; set; } // DateTimeOffset

        [JsonProperty("location")]
        public string Location { get; set; }

        [JsonProperty("geo_enabled")]
        public bool? GeoEnabled { get; set; }

        [JsonProperty("url")]
        public string Url { get; set; }

        [JsonProperty("statuses_count")]
        public int StatusesCount { get; set; }

        [JsonProperty("followers_count")]
        public int FollowersCount { get; set; }

        [JsonProperty("friends_count")]
        public int FriendsCount { get; set; }

        [JsonProperty("following")]
        public bool? Following { get; set; }

        public bool Protected { get; set; }

        [JsonProperty("verified")]
        public bool Verified { get; set; }

        //[JsonProperty("entities")]
        //public IUserEntities Entities { get; set; }

        [JsonProperty("notifications")]
        public bool? Notifications { get; set; }

        //[JsonProperty("profile_image_url")]
         // public string ProfileImageUrlHttp { get; set; }

        [JsonProperty("profile_image_url_https")]
        public string ProfileImageUrlHttps { get; set; }

        [JsonProperty("follow_request_sent")]
        public bool? FollowRequestSent { get; set; }

        [JsonProperty("default_profile")]
        public bool DefaultProfile { get; set; }

        [JsonProperty("default_profile_image")]
        public bool DefaultProfileImage { get; set; }

        [JsonProperty("favourites_count")]
        public int FavoritesCount { get; set; }

        //[JsonProperty("listed_count")]
        public int ListedCount { get; set; }

        [JsonProperty("profile_sidebar_fill_color")]
         public string ProfileSidebarFillColor { get; set; }

        [JsonProperty("profile_sidebar_border_color")]
        public string ProfileSidebarBorderColor { get; set; }

        [JsonProperty("profile_background_tile")]
        public bool ProfileBackgroundTile { get; set; }

        [JsonProperty("profile_background_color")]
        public string ProfileBackgroundColor { get; set; }

        [JsonProperty("profile_background_image_url")]
        public string ProfileBackgroundImageUrl { get; set; }

        [JsonProperty("profile_background_image_url_https")]
        public string ProfileBackgroundImageUrlHttps { get; set; }

        [JsonProperty("profile_banner_url")]
        public string ProfileBannerURL { get; set; }

        [JsonProperty("profile_text_color")]
        public string ProfileTextColor { get; set; }

        [JsonProperty("profile_link_color")]
        public string ProfileLinkColor { get; set; }

        [JsonProperty("profile_use_background_image")]
        public bool ProfileUseBackgroundImage { get; set; }

        [JsonProperty("is_translator")]
        public bool IsTranslator { get; set; }

        [JsonProperty("contributors_enabled")]
        public bool ContributorsEnabled { get; set; }

        [JsonProperty("utc_offset")]
        public int? UtcOffset { get; set; }

        [JsonProperty("time_zone")]
        public string TimeZone { get; set; }

        //[JsonProperty("withheld_in_countries")]
        //public IEnumerable<string> WithheldInCountries { get; set; }

        //[JsonProperty("withheld_scope")]
        //public string WithheldScope { get; set; }

        public override string ToString()
        {
            return base.ScreenName;
        }
    }
}
